bigframes/core/compile/polars/compiler.py

```diff
@@ -0,0 +1,373 @@
+        if not polars_installed:
+            raise ValueError(
+                "Polars is not installed, cannot compile to polars engine."
+            )
+
+        # Polars has incomplete slice support in lazy mode
+        node = bigframes.core.rewrite.replace_slice_ops(array_value.node)
+        return self.compile_node(node)
+
+    @functools.singledispatchmethod
+    def compile_node(self, node: nodes.BigFrameNode):
+        """Defines transformation but isn't cached, always use compile_node instead"""
+        raise ValueError(f"Can't compile unrecognized node: {node}")
+
+    @compile_node.register
+    def compile_readlocal(self, node: nodes.ReadLocalNode):
+        return pl.read_ipc(node.feather_bytes).lazy()
+
+    @compile_node.register
+    def compile_filter(self, node: nodes.FilterNode):
+        return self.compile_node(node.child).filter(
+            self.expr_compiler.compile_expression(node.predicate)
+        )
+
+    @compile_node.register
+    def compile_orderby(self, node: nodes.OrderByNode):
+        frame = self.compile_node(node.child)
+        if len(node.by) == 0:
+            # pragma: no cover
+            return frame
+
+        frame = frame.sort(
+            [
+                self.expr_compiler.compile_expression(by.scalar_expression)
+                for by in node.by
+            ],
+            descending=[not by.direction.is_ascending for by in node.by],
+            nulls_last=[by.na_last for by in node.by],
+            maintain_order=True,
+        )
+        return frame
+
+    @compile_node.register
+    def compile_reversed(self, node: nodes.ReversedNode):
+        return self.compile_node(node.child).reverse()
+
+    @compile_node.register
+    def compile_selection(self, node: nodes.SelectionNode):
+        return self.compile_node(node.child).select(
+            **{new.sql: orig.id.sql for orig, new in node.input_output_pairs}
+        )
+
+    @compile_node.register
+    def compile_projection(self, node: nodes.ProjectionNode):
+        new_cols = [
+            self.expr_compiler.compile_expression(ex).alias(name.sql)
+            for ex, name in node.assignments
+        ]
+        return self.compile_node(node.child).with_columns(new_cols)
+
+    @compile_node.register
+    def compile_rowcount(self, node: nodes.RowCountNode):
+        rows = self.compile_node(node.child).count()[0]
+        return pl.DataFrame({node.col_id.sql, [rows]})
+
+    @compile_node.register
+    def compile_offsets(self, node: nodes.PromoteOffsetsNode):
+        return self.compile_node(node.child).with_columns(
+            [pl.int_range(pl.len(), dtype=pl.Int64).alias(node.col_id.sql)]
+        )
+
+    @compile_node.register
+    def compile_slice(self, node: nodes.SliceNode):
+        return self.compile_node(node.child)[node.start : node.stop : node.step]
+
+    @compile_node.register
+    def compile_join(self, node: nodes.JoinNode):
```

Comment:
Maybe add a comment saying we always do ordered mode locally since it's easy with polars?

Reply:
added